docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,251 @@
+* The existing unbind API will be used to trigger this process
+  * This means removing the cluster ref from the agent CR when using the kubeAPI
+
+#### Example boot entry
+
+```
+$ cat /boot/loader/entries/00-assisted-discovery.conf
+title Assisted Installer Discovery
+version 999
+options random.trust_cpu=on ignition.firstboot ignition.platform.id=metal 'coreos.live.rootfs_url=http://10.1.155.24:6016/boot-artifacts/rootfs?arch=x86_64&version=4.10'
+linux /boot/discovery/vmlinuz
+initrd /boot/discovery/c4bd9751-f3ae-49ac-a36b-d3ffebef4a39-initrd.img
+```
+
+The name of this file and the version number ensure that it should always be the
+first boot entry and will be selected upon reboot.
+
+#### State Machine Changes
+Currently when a host is installed, all unbind APIs will call the `UnbindHost`
+transition. This will always move the host to
+`HostStatusUnbindingPendingUserAction`. If the agent has a BMH associated with
+it, the bare metal host agent controller (BMAC) will make the changes required
+to boot the host back into discovery. If there is no BMH associated with the
+agent (or the kube API is not in use), the host will remain in
+`HostStatusUnbindingPendingUserAction`.
+
+The following diagram illustrates the new proposed transition to indicate that a
+running agent should execute the reclaim procedure.
+
+```
+                   -----------------------
+                   | HostStatusInstalled |
+                   -----------------------
+                             |
+                      User unbinds agent
+                             |
+     Kube API and no BMH     |   REST API, Kube API with BMH, or other machine management
+          -------------------+-------------------
+          |                                     |
+          V                                     V
+-----------------------------            ----------------------------
+| TransitionTypeReclaimHost |            | TransitionTypeUnbindHost |
+-----------------------------            ----------------------------
+            |                                               |
+            V                                               |
+  ------------------------                                  |
+  | HostStatusReclaiming |                                  |
+  ------------------------                                  |
+                |                                           |
+       Agent posts step reply                               |
+                |                                           |
+       ---------+----------------------------------         |
+       |                                          |         |
+    Success                                    Failure      |
+       |                                          |         |
+       V                                          |         |
+-------------------------------------             |         |
+| HostStatusReclaimingPendingReboot |------       |         |
```

Comment:
I think we changed this to ReclaimingRebooting?

Reply:
Yup, let me update.